src/hardware/voodoo.cpp

```diff
@@ -0,0 +1,7175 @@
+}
+
+inline UINT8 count_leading_zeros(UINT32 value)
+{
+#ifdef _MSC_VER
+	DWORD idx = 0;
+	return (_BitScanReverse(&idx, value) ? (UINT8)(31 - idx) : (UINT8)32);
+#else
+	return (value ? (UINT8)__builtin_clz(value) : (UINT8)32);
+#endif
+	//INT32 result;
+	//
+	//#if defined _MSC_VER
+	//__asm
+	//{
+	//	bsr   eax,value
+	//	jnz   skip
+	//	mov   eax,63
+	//skip:
+	//	xor   eax,31
+	//	mov   result,eax
+	//}
+	//#else
+	//result = 32;
+	//while(value > 0) {
+	//	result--;
+	//	value >>= 1;
+	//}
+	//#endif
+	//return (UINT8)result;
+}
+
+/*************************************
+ *
+ *  Computes a fast 16.16 reciprocal
+ *  of a 16.32 value; used for
+ *  computing 1/w in the rasterizer.
+ *
+ *  Since it is trivial to also
+ *  compute log2(1/w) = -log2(w) at
+ *  the same time, we do that as well
+ *  to 16.8 precision for LOD
+ *  calculations.
+ *
+ *  On a Pentium M, this routine is
+ *  20% faster than a 64-bit integer
+ *  divide and also produces the log
+ *  for free.
+ *
+ *************************************/
+
+inline INT64 fast_reciplog(INT64 value, INT32* log2)
+{
+	UINT32 temp, rlog;
+	UINT32 interp;
+	UINT32 *table;
+	UINT64 recip;
+	bool neg = false;
+	int lz, exp = 0;
+
+	/* always work with unsigned numbers */
+	if (value < 0)
+	{
+		value = -value;
+		neg = true;
+	}
+
+	/* if we've spilled out of 32 bits, push it down under 32 */
+	if (value & LONGTYPE(0xffff00000000))
+	{
+		temp = (UINT32)(value >> 16);
+		exp -= 16;
+	}
+	else
+		temp = (UINT32)value;
+
+	/* if the resulting value is 0, the reciprocal is infinite */
+	if (GCC_UNLIKELY(temp == 0))
+	{
+		*log2 = 1000 << LOG_OUTPUT_PREC;
+		return neg ? 0x80000000 : 0x7fffffff;
+	}
+
+	/* determine how many leading zeros in the value and shift it up high */
+	lz = count_leading_zeros(temp);
+	temp <<= lz;
+	exp += lz;
+
+	/* compute a pointer to the table entries we want */
+	/* math is a bit funny here because we shift one less than we need to in order */
+	/* to account for the fact that there are two UINT32's per table entry */
+	table = &voodoo_reciplog[(temp >> (31 - RECIPLOG_LOOKUP_BITS - 1)) & ((2 << RECIPLOG_LOOKUP_BITS) - 2)];
+
+	/* compute the interpolation value */
+	interp = (temp >> (31 - RECIPLOG_LOOKUP_BITS - 8)) & 0xff;
+
+	/* do a linear interpolatation between the two nearest table values */
+	/* for both the log and the reciprocal */
+	rlog = (table[1] * (0x100 - interp) + table[3] * interp) >> 8;
+	recip = (table[0] * (0x100 - interp) + table[2] * interp) >> 8;
+
+	/* the log result is the fractional part of the log; round it to the output precision */
+	rlog = (rlog + (1 << (RECIPLOG_LOOKUP_PREC - LOG_OUTPUT_PREC - 1))) >> (RECIPLOG_LOOKUP_PREC - LOG_OUTPUT_PREC);
+
+	/* the exponent is the non-fractional part of the log; normally, we would subtract it from rlog */
+	/* but since we want the log(1/value) = -log(value), we subtract rlog from the exponent */
+	*log2 = ((exp - (31 - RECIPLOG_INPUT_PREC)) << LOG_OUTPUT_PREC) - rlog;
```

Comment:
If `31 - RECIPLOG_INPUT_PREC > exp`, we'll be left-shifting a negative value. Maybe `<< LOG_OUTPUT_PREC` should just multiply by 256 instead?